packages/babel-traverse/src/path/inference/inferers.js

```diff
@@ -83,17 +83,29 @@ export function BinaryExpression(node) {
 }
 
 export function LogicalExpression() {
-  return t.createUnionTypeAnnotation([
+  const argumentTypes = [
     this.get("left").getTypeAnnotation(),
     this.get("right").getTypeAnnotation(),
-  ]);
+  ];
+
+  if (argumentTypes.every(t.isTSTypeAnnotation)) {
+    return t.createTSUnionType(argumentTypes);
+  }
+
+  return t.createUnionTypeAnnotation(argumentTypes);
 }
 
 export function ConditionalExpression() {
-  return t.createUnionTypeAnnotation([
+  const argumentTypes = [
     this.get("consequent").getTypeAnnotation(),
     this.get("alternate").getTypeAnnotation(),
-  ]);
+  ];
+
+  if (argumentTypes.every(t.isTSTypeAnnotation)) {
```

Comment:
The fix 💊

Changes are also applied for
* `LogicalExpression` - https://github.com/babel/babel/pull/11378/files#diff-6c956a118dc9996dff6b731e42856479R85
* `getTypeAnnotationBindingConstantViolations` - https://github.com/babel/babel/pull/11378/files#diff-1fce3c9f392392fe193c399ad4a97b65R99
* `getConditionalAnnotation` - https://github.com/babel/babel/pull/11378/files#diff-1fce3c9f392392fe193c399ad4a97b65R212